geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/KryoLazyDensityIterator.scala

```diff
@@ -0,0 +1,251 @@
+    }
+  }
+
+  override def newResult() = mutable.Map.empty[(Int, Int), Double]
+
+  override def aggregateResult(sf: SimpleFeature, result: DensityResult): Unit = writeGeom(sf, result)
+
+  override def encodeResult(result: DensityResult): Array[Byte] = KryoLazyDensityIterator.encodeResult(result)
+
+  /**
+   * Gets the weight for a feature from a double attribute
+   */
+  private def getWeightFromDouble(i: Int)(sf: SimpleFeature): Double = {
+    val d = sf.getAttribute(i).asInstanceOf[java.lang.Double]
+    if (d == null) 0.0 else d
+  }
+
+  /**
+   * Tries to convert a non-double attribute into a double
+   */
+  private def getWeightFromNonDouble(i: Int)(sf: SimpleFeature): Double = {
+    val d = sf.getAttribute(i)
+    if (d == null) {
+      0.0
+    } else {
+      val converted = Converters.convert(d, classOf[java.lang.Double])
+      if (converted == null) 1.0 else converted
+    }
+  }
+
+  /**
+   * Evaluates an arbitrary expression against the simple feature to return a weight
+   */
+  private def getWeightFromExpression(e: Expression)(sf: SimpleFeature): Double = {
+    val d = e.evaluate(sf, classOf[java.lang.Double])
+    if (d == null) 0.0 else d
+  }
+
+  /**
+   * Writes a density record from a feature that has a point geometry
+   */
+  def writePoint(sf: SimpleFeature, weight: Double, result: DensityResult): Unit =
+    writePointToResult(sf.getAttribute(geomIndex).asInstanceOf[Point], weight, result)
+
+  /**
+   * Writes a density record from a feature that has an arbitrary geometry
+   */
+  def writeNonPoint(geom: Geometry, weight: Double, result: DensityResult): Unit =
+    writePointToResult(geom.getCentroid, weight, result)
+
+  protected[iterators] def writePointToResult(pt: Point, weight: Double, result: DensityResult): Unit =
+    writeSnappedPoint((gridSnap.i(pt.getX), gridSnap.j(pt.getY)), weight, result)
+
+  protected[iterators] def writePointToResult(pt: Coordinate, weight: Double, result: DensityResult): Unit =
+    writeSnappedPoint((gridSnap.i(pt.x), gridSnap.j(pt.y)), weight, result)
+
+  protected[iterators] def writePointToResult(x: Double, y: Double, weight: Double, result: DensityResult): Unit =
+    writeSnappedPoint((gridSnap.i(x), gridSnap.j(y)), weight, result)
+
+  private def writeSnappedPoint(xy: (Int, Int), weight: Double, result: DensityResult): Unit =
+    result.update(xy, result.getOrElse(xy, 0.0) + weight)
+
+  override def deepCopy(env: IteratorEnvironment): SortedKeyValueIterator[Key, Value] = ???
+}
+
+object KryoLazyDensityIterator extends Logging {
+
+  type DensityResult = mutable.Map[(Int, Int), Double]
+  type GridIterator = (SimpleFeature) => Iterator[(Double, Double, Double)]
+
+  // needs to be lazy to avoid class loading issues before init is called
+  lazy val DENSITY_SFT = SimpleFeatureTypes.createType("density", "result:String,*geom:Point:srid=4326")
+
+  val DEFAULT_PRIORITY = 25
```

Comment:
Yeah, we originally had these in a common place... I personally think it's easier to keep track of when it's with the iterator. In practice, we almost never combine more than 1 iterator, so the priority is semi-irrelevant. (the caveat to that is that accumulo system iterators run at priority 15 or something, and we don't want to run before those).
It would be nice to have a better way to track them all though...